Source Code Version Control Events section

salaboy opened the pull request:
PR to discuss Source code version control Events

m-linner-ericsson commented on vocabulary-draft/source-code-version-control.md:
```diff
@@ -0,0 +1,27 @@
+- **Repository Deleted Event**: a Source Code Repository was deleted and it is not longer available
+- **Branch Created Event**: a Branch inside the Repository was created 
+- **Branch Deleted Event**: a Branch inside the Repository was deleted
+- **Change Committed**: a change is commited to a branch
```
When using Gerrit we could have the following scenario:

- Write your code, commit it and push your changes to the refs/for/<branch name> (New change-id)
- Review of the code
- Update to fix the comments (new commit)
- Approve the change and merge it

When would we send the **Change Committed** event? At each push or one event for the whole change?

salaboy replied:
@m-linner-ericsson this is up for discussion, but the original intent that I was trying to capture is when you push a commit (a batch of changes) to a branch.

e-backmark-ericsson replied:
@salaboy , I guess you mean the action that in GitHub and Gitlab is called _merge_ and in Gerrit is called _submit_ (as stated here: https://github.com/cdfoundation/sig-interoperability/blob/master/docs/vocabulary.md#scm-tools-and-technologies) or? This would be the natural event to kick off a CI pipeline on.
@m-linner-ericsson , I think you're requesting an event stating that a new commit has been added for review (called patch-set in Gerrit and PR/MR commit in GitHub/Gitlab), right? In Ericsson we often kick off _pre-merge_ pipelines based on such events, to automatically run any tests on the commit prior to merging/submitting it.

erkist replied:
My comment from the old PR:
> For me, the "Change being proposed as in a Pull Request" would correspond to Change Created (and Change Updated if new commits are added to the PR), and "Change being merged to a branch" would correspond to Change Merged.

I should clarify for Change Merged that I mean merged to the target branch for the review, not to the "review branch"

salaboy replied:
This is super useful feedback, I think that now we are close to defining the terms that we need, I am not strongly attached to any of those terms, I think that choosing one that is widely recognized in the industry will make it easier to adopt that brand new terms defined by us, so I will tend to go for GitHub/Gitlab terms. Do you have any preferences?

m-linner-ericsson commented on vocabulary-draft/source-code-version-control.md:
```diff
@@ -0,0 +1,27 @@
+- **Branch Created Event**: a Branch inside the Repository was created 
+- **Branch Deleted Event**: a Branch inside the Repository was deleted
+- **Change Committed**: a change is commited to a branch
+- **Change Approved**: a change request is created by a user, that requires review and approval before being applied to an existing branch inside the repository
```
If a review requires several approvals would we send an event for each? For example

- Code review 
- CI engine verifications
- Code guardian approval

salaboy replied:
I think that `Change Approved` was trying to reflect that. We can capture different actors for example the `Approver` with data inside the event itself. Again I was trying to focus on the minimal set of metadata that we need to get started. But we can definitely add more to it.

m-linner-ericsson replied:
So your idea was to send one event when all approvers have given their content? If we go with one event for each approval we could:

* Capture code +1 but CI engine -1
* Know the time between submission and guardian approval

Or was there some other idea?

salaboy replied:
@m-linner-ericsson we can send an event per approver and another event when all the expected approvers are done. I think that we can use an attribute to identify if these approvers are people or automated software.

m-linner-ericsson commented on vocabulary-draft/source-code-version-control.md:
```diff
@@ -0,0 +1,27 @@
+- **Repository URL**: indicates the location of the source code repository for API operations, this URL needs to include the protocol used to connect to the repository. For example `git://` , `ssh://`, `https://`
+- **Repository Name**: friendly name to list this repository to users
+
+Optional attributes: 
```
Do we see a need to be able to link directly to a PR or commit in an event?

salaboy replied:
@m-linner-ericsson I think that we need to link to both, as optional metadata

e-backmark-ericsson commented:
I think it would be beneficial to differ between a _source change_ and a _commit_, where a _source change_ would correspond to a GitHub Pull Request, Gitlab Merge Request or Gerrit Change and a _commit_ obviously would be an individual commit within such a change. With that I would propose something like these event types:
- Source Change Created (corresponding to a new PR/MR/Change created)
- Source Change Approved (corresponding to the Approved state in Gitlab for example)
- Source Change Merged (corresponding to a PR/MR/Change merged)
- Source Changed Abandoned (corresponding to an abandoned PR/MR/Change)
- Commit Created (corresponding to a new commit added to an existing source change)
I see that it could be interesting to be able to trigger pipeline runs on most of those events, and also to calculate KPIs and base visualizations on them

erkist commented on vocabulary-draft/source-code-version-control.md:
```diff
@@ -0,0 +1,27 @@
+
+Optional attributes: 
+- **Repository Owner**: indicates who is the owner of the repository, usually a `user` or an `organization`
+- **Repository URL View**: URL to access the repository from a user web browser
```
I would prefer "View URL" over "URL View", as it is the URL to the view, not the view of the URL.

salaboy replied:
done

salaboy commented:
@e-backmark-ericsson I do like your suggestions, let me add that to this PR and lets review again. I do really like the progress we are making here.. now it feels more concrete

salaboy commented:
@e-backmark-ericsson based on your suggestions for the Change Events I've modified that section to try to cover your and others suggestions. 
The only change from my side on top of that were: 
- Removed **Source** from Change events as I think that we are talking about source code repositories, so I included source code change in the definition of the events instead of the name. I wanted to keep the name as concise as possible, I am happy to change it if someone thinks that Source Change is more clear and it really makes a difference
- I've changed Commit Created to Commit Added, as when I was trying to define what it means, this commit here is an addition to a Change and not a random commit happening somewhere else.

m-linner-ericsson commented on vocabulary-draft/source-code-version-control.md:
```diff
@@ -0,0 +1,36 @@
+- **Change Approved Event**:  someone (user) or an automated system submitted an approval to the source code change. A user or an automated system needs to be in charge of understanding how many approvals are needed for this change to be merged.    
+- **Change Merged Event**: the change is merged to the target branch where it was submitted. 
+- **Change Abandoned Event**: a tool or a user decides that the change has been inactive for a while and it can be considered abandoned.
+- **Commit Added**: a new commit is added to an existing Change
```
Should this be called **Change updated** or something similar to indicate that it is connected to the other **Change** events. I am afraid we will confuse people here.

salaboy replied:
@m-linner-ericsson yeah.. I was thinking the same.. let's change that for now.

erkist commented on vocabulary-draft/source-code-version-control.md:
```diff
@@ -0,0 +1,36 @@
+
+These events are related to Source Code repositories
+- **Repository Created Event**: a new Source Code Repository was created to host source code for a project
+- **Repository Modified Event**: a Source Code Repository modified some of its attributes, like location, or owner
```
I think there is a risk of confusion between the modification of repository "information" (e.g. location or owner) and modification of the repository "content", e.g. new branches or commits.

I propose we rename this event to either `repository metadata updated` or `repository information updated`

salaboy replied:
@erkist so to be clear, you think that the name of the event needs to be more clear. In my perspective, the description is stating that already.  I wonder if for example the repository URL is considered metadata, as where the repository represents the repository itself. I am happy to change the name of the event if that makes it more clear.

erkist replied:
@salaboy I agree that the description makes it clear, but would prefer if it was clear already by the name of the event.
And maybe metadata is the wrong term? I'm looking for a term that indicates that it is not the content of the repository that has been modified (see for instance https://git-scm.com/book/en/v2/Git-Basics-Recording-Changes-to-the-Repository which uses the term modification to talk about the content of the repo).

But I don't have a strong opinion on it.

erkist commented on vocabulary-draft/source-code-version-control.md:
```diff
@@ -0,0 +1,36 @@
+- **Repository Created Event**: a new Source Code Repository was created to host source code for a project
+- **Repository Modified Event**: a Source Code Repository modified some of its attributes, like location, or owner
+- **Repository Deleted Event**: a Source Code Repository was deleted and it is not longer available
+- **Branch Created Event**: a Branch inside the Repository was created 
```
Should we have some event for announcing that a branch has been modified, i.e. it now points to a different commit?

salaboy replied:
@erkist that is a good example of something that I think is extremely useful but it might not be necessary to include the initial set of events. How do we document these things to not loose track of them, but at the same time don't push for all these events to be present in the first iteration of the vocabulary (to reduce mental overhead)

erkist commented on vocabulary-draft/source-code-version-control.md:
```diff
@@ -0,0 +1,36 @@
+
+From each repository you can emit events related with proposed source code changes. Each change can include a single or multiple commits that can also be tracked. 
+
+- **Change Submitted Event**: a source code change was created and submitted to a repository specific branch. Examples: PullRequest sent to Github, MergeRequest sent to Gitlab, Change created in Gerrit
```
As mentioned before, this is quite confusing for Gerrit users, as the operation to merge a change is called "Submit". Can we call this Change Created instead? Or what does the "submitted" part of "created and submitted" mean?

salaboy replied:
@erkist yes.. I think that Change Created is clear enough.. as the change is created into a repository, hence is already submitted.

salaboy replied:
changed

erkist commented on vocabulary-draft/source-code-version-control.md:
```diff
@@ -0,0 +1,36 @@
+From each repository you can emit events related with proposed source code changes. Each change can include a single or multiple commits that can also be tracked. 
+
+- **Change Submitted Event**: a source code change was created and submitted to a repository specific branch. Examples: PullRequest sent to Github, MergeRequest sent to Gitlab, Change created in Gerrit
+- **Change Approved Event**:  someone (user) or an automated system submitted an approval to the source code change. A user or an automated system needs to be in charge of understanding how many approvals are needed for this change to be merged.    
```
I guess this would also apply to "negative approvals", e.g. a reviewer gives a negative score, or an automated check fails?

If so, I'd propose either "Change Vote Event" or "Change Approval Changed Event" or similar, as the event may just as well signal that the change is **not** approved.

salaboy commented on vocabulary-draft/source-code-version-control.md:
```diff
@@ -0,0 +1,37 @@
+From each repository you can emit events related with proposed source code changes. Each change can include a single or multiple commits that can also be tracked. 
+
+- **Change Created Event**: a source code change was created and submitted to a repository specific branch. Examples: PullRequest sent to Github, MergeRequest sent to Gitlab, Change created in Gerrit
+- **Change Reviewed Event**:  someone (user) or an automated system submitted an review to the source code change. A user or an automated system needs to be in charge of understanding how many approvals/rejections are needed for this change to be merged or rejected. The review event needs to include if the change is approved by the reviewer, more changes are needed or if the change is rejected.   ...
```
@erkist I've updated this event to be Change Reviewed event, including positive reviews and negative reviews. Here in github you have those 3 outcomes, approved, more changes requested and rejected as far as I remember.

afrittoli commented on vocabulary-draft/source-code-version-control.md:
```diff
@@ -0,0 +1,37 @@
+
+
+Repository Events MUST include the following attributes:
+- **Event Type**: the type is restricted to include `CD.Repository**` prefix. For example `CD.Repository.Created` or `CD.Repository.ChangeApproved`
```
NIT: can use lowercase only for the event types?

salaboy replied:
yes.. I was thinking about it.. let me do those changes now

afrittoli commented on vocabulary-draft/source-code-version-control.md:
```diff
@@ -0,0 +1,37 @@
+- **Change Created Event**: a source code change was created and submitted to a repository specific branch. Examples: PullRequest sent to Github, MergeRequest sent to Gitlab, Change created in Gerrit
+- **Change Reviewed Event**:  someone (user) or an automated system submitted an review to the source code change. A user or an automated system needs to be in charge of understanding how many approvals/rejections are needed for this change to be merged or rejected. The review event needs to include if the change is approved by the reviewer, more changes are needed or if the change is rejected.   ...
+- **Change Merged Event**: the change is merged to the target branch where it was submitted. 
+- **Change Abandoned Event**: a tool or a user decides that the change has been inactive for a while and it can be considered abandoned.
```
I think "abandoned" here is a specific case of closed. You can have change explicitly closed instead of abandoned. 
One could argue that change "merged" is also a specific case of "closed". 
Perhaps we could have a "closed" event with a reason field, which can me "merged" or "closed without reason" or "expired" etc.

afrittoli commented on vocabulary-draft/source-code-version-control.md:
```diff
@@ -0,0 +1,37 @@
+
+From each repository you can emit events related with proposed source code changes. Each change can include a single or multiple commits that can also be tracked. 
+
+- **Change Created Event**: a source code change was created and submitted to a repository specific branch. Examples: PullRequest sent to Github, MergeRequest sent to Gitlab, Change created in Gerrit
```
A **change** is an abstraction created by tools like gerrit, github and gitlab on top of `git`. 
We need to have **change** specific events, but I wonder if we should have events for the underlying repo too.
For instance when a **change** is closed, you could from zero to many commits added to the repo. 
Do we want separate events for that, or do we want to embed that information in the change event? 
You can have commits added to a repo without a change too, as most of this systems allow direct interactions with the underlying source code version control system. 

The data model can easily explode in this space (see all github events for instance), so I think it's fine if we can it simple in the beginning, but I think it's important to make sure we have an event when something changes in the repository.

salaboy replied:
@afrittoli I tend to agree.. let's keep it simple for now, but let's make sure that we discuss how can we tackle more granular changes. We need to have a clear way to make decisions about for example github events, what situations to cover and which ones are not strictly relevant. For making those decisions I propose to define a set of scenarios (and PoCs) that we want to cover with the first iteration and then expand from there.
